### PR TITLE
Using a spinlock worker instead of a thread per batch

### PR DESCRIPTION
**Pull request by lemire**
This is an alternative design to PR https://github.com/simdjson/simdjson/pull/920 which relies on a more conventional thread worker for accelerating `parse_many`.

This PR:
```
$ ./benchmark/parse_stream bar.ndjson
parse_many: Speed per batch_size... from 10000 bytes to 10000000 bytes...
Batch Size	Gigabytes/second	Nb of documents parsed
10000		3.517				793000
109900		3.651				793000
209800		3.690				793000
309700		3.695				793000
409600		3.611				793000
509500		3.556				793000
609400		3.581				793000
709300		3.574				793000
809200		3.558				793000
909100		3.555				793000
1009000		3.552				793000
Seemingly optimal batch_size: 309700...
Best speed: 3.69506...
```

PR 920:

```
$ ../build/benchmark/parse_stream bar.ndjson
parse_many: Speed per batch_size... from 10000 bytes to 10000000 bytes...
Batch Size	Gigabytes/second	Nb of documents parsed
10000		1.368				793000
109900		3.531				793000
209800		3.611				793000
309700		3.640				793000
409600		3.639				793000
509500		3.554				793000
609400		3.564				793000
709300		3.557				793000
809200		3.551				793000
909100		3.551				793000
1009000		3.551				793000
Seemingly optimal batch_size: 309700...
Best speed: 3.63969...
```

The main point is that this new PR will be fast over a broad range of batch sizes because there is minimal overhead due to thread-to-thread communication. The PR https://github.com/simdjson/simdjson/pull/920  is better than master in that it does not create a whole new thread with each window, but it still not optimally fast because it relies on the operating system to wake up the worker thread.

The negative of this new PR is that it relies on spinlocks which are fast, but they burn CPU cycles while waiting. This is not a problem if they are only meant to waiting a short time. It is not like the OS is likely to be able to squeeze in much work in a few microseconds here and there.

So what is the difference? In my tests I get a 15%--20% total increase in the recorded CPU usage due (presumably) to the spinlock. This is a rough measure obtained from running `perf stat`.

Fixes https://github.com/simdjson/simdjson/issues/529

### CODE REVIEW

**Comment by lemire** [comment]
The failing tests are related to https://github.com/simdjson/simdjson/pull/926

File: include/simdjson/inline/document_stream.h [hunk 1]
```diff
@@ -5,10 +5,62 @@
 #include <algorithm>
 #include <limits>
 #include <stdexcept>
-
 namespace simdjson {
 namespace dom {
 
+#ifdef SIMDJSON_THREADS_ENABLED
+inline void stage1_worker::finish() {
```
[the remark this answers is not in the document]
Comment (jkeiser) [reply include/simdjson/inline/document_stream.h:12]: So the issue was that the OS took forever to wake the thread?

**Comment by jkeiser** [comment]
It's hard to decide. This one is faster for all batch sizes, but we could mitigate batch size issues that with docs or warnings. Stressing the CPU less is a really big deal; 15-20% extra stress every other thread translates to a 7-10% reduced throughput on a server. For a lot of people, the reason they want the speed because lets them handle more requests, and the actual per-request speed is gravy.

Also, if the spinlock is really the reason it's worse, wouldn't that mean we see bigger and bigger CPU usage the larger your batch size gets? After all the thread has longer to wait in those cases ...

I guess if I had to choose I'd go with the thread wake patch. Partly because I'm focused on server use cases.

I imagine there will be ways to reduce or mitigate either version, though.

**Comment by lemire** [comment]
> I guess if I had to choose I'd go with the thread wake patch. Partly because I'm focused on server use cases.

I agree. I am uneasy with the spinlock anyhow.

**Comment by lemire** [comment]
I am retiring this PR"

**Pull request closed without merging**